### PR TITLE
[AIR][Good for initial review] Add numpy-only path for DLPredictor 

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

There're a few problems we observed from recent demo CUJ:
* Our predictor API supports user to subclass and implement their own `call_model()` to produce outputs, but only support `Tensor` or `Dict[str, Tensor]`. This means for all other output formats use needs to figure out how to manually pad and concat tensors themselves ... like this
```python
class VGG16Predictor(TorchPredictor):
    def call_model(self, tensor):
        model_output = super().call_model(tensor)

        # Find out the largest number of boxes, any of these
        # images have.
        pad_dim = max([output["boxes"].shape[0] for output in model_output])

        objects = {}
        for obj in model_output:
            for k, v in obj.items():
                if k not in objects: objects[k] = []
                v = v.detach().cpu().numpy()
                # Potentially pad the data for this column to the max length.
                pad = np.zeros((pad_dim - v.shape[0],) + v.shape[1:])
                v = np.concatenate((v, pad), axis=0)
                # Append so we can batch later.
                objects[k].append(v)

        for k, v in objects.items():
            objects[k] = torch.tensor(np.array(v))

        return objects
```
* For DLPredictor types, we still enforces pd.DataFrame in multiple places that incurs at least two extra copies for each request. This hurts perf, and also leaks pdands abstraction to user that user's current best UX is to subclass and implement their own predictor and wrap in pandas:

```python
class SSDPredictor(TorchPredictor):
    def call_model(
        self, tensor: Union[torch.Tensor, Dict[str, torch.Tensor]]
    ) -> Union[torch.Tensor, Dict[str, torch.Tensor], pd.DataFrame]:
        """User predictor output formatting code."""
        model_output = super().call_model(tensor)
        return pd.DataFrame([
            {k: v.detach().cpu().numpy() for k, v in objects.items()}
            for objects in model_output
        ])
```

## What this PR does

* Add a tensor/numpy narrow waist path for preprocessing and predictor.
  * User can write tensor centric code for preprocessor and predictor UDFs
  * No need to write custom predictor in most cases, supports various output types and ragged tensor by default 
  * Better perf by eliminating pandas <> numpy copies for each prediction request.
* Now there will be only one casting to pandas at the final step, where we can decide with try-catching special exception type to allow user getting a pd.DataFrame of free-form prediction output data. The assumption here is outputs can be very heterogenous but meanwhile data size is small & not strong requirement of linking output to other downstream components, thus `TensorArray` related optimizations and restrictions are not strongly related.

As a result, with this PR, both code snippets will just be eliminated from user's critical path, just using plain `TorchPredictor` works out of the box.

## Next steps 

Given we're close to Ray 2.0 rc1, we need to find the safest and least risky path to get benefits without significantly change how things are done on master.

My personal recommendation for next steps are..
* We limit this tensor narrow waist only to DLPredictors to reduce surface area
* We add a feature flag just like the auto tensor array casting, but default as False (disable) until we're certain it won't cause regression.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/preprocessor.py
Comment (clarkzinzow) [inline python/ray/data/preprocessor.py:290]: I think we'd want to do the same conversion semantics as our data batch conversion, i.e. a single-column table is turned into a single ndarray, while a multi-column table is turned into a dict of ndarrays: https://github.com/ray-project/ray/blob/ef260702a20c8424ebc7914bf72800e5c409fb29/python/ray/air/util/data_batch_conversion.py#L86-L95

File: python/ray/data/preprocessor.py
Comment (clarkzinzow) [inline python/ray/data/preprocessor.py:303]: I _think_ that we would want to support the full NumPy narrow waist here, including `Dict[str, np.ndarray]` for multi-column tables.
```suggestion
    def _transform_numpy(self, np_array: Union["np.ndarray", Dict[str, "np.ndarray"]]) -> Union["np.ndarray", Dict[str, "np.ndarray"]]:
```

File: python/ray/data/preprocessor.py
Comment (clarkzinzow) [inline python/ray/data/preprocessor.py:11]: Small nit: NumPy is available as a Ray dependency so we don't need to guard it under `TYPE_CHECKING`.

File: python/ray/train/batch_predictor.py
Comment (clarkzinzow) [inline python/ray/train/batch_predictor.py:182]: Hmm is this needed/desired? I think that we'd still want to try to cast the tensor model outputs to our extension type, with a fallback?

If we are not going to do a transparent fallback, I don't think that we should mutate the context here (which would affect other future computations launched by this driver); we could instead just override `cast_tensor_columns` which is used on the model output below.

File: python/ray/train/batch_predictor.py
Comment (clarkzinzow) [inline python/ray/train/batch_predictor.py:201]: `batch` may be a single ndarray for the `"numpy"` batch format, so that case should be handled as well.

File: python/ray/train/batch_predictor.py
Comment (clarkzinzow) [inline python/ray/train/batch_predictor.py:220]: We could technically returned `prediction_output` uncasted, and Datasets will construct an Arrow Table from the ndarray or dict of ndarrays, but unlike the Pandas block representation, this would not be able to represent ragged tensors... I'm thinking that maybe the try-except fallback on the tensor extension casting makes sense here?

**Review by clarkzinzow — COMMENTED** [review]
Awesome work, I'm happy that this ended up working out ~how we expected!

File: python/ray/data/preprocessor.py
Comment (jiaodong) [reply python/ray/data/preprocessor.py:303]: yep you're totally right on this, i wrote this signature first and only realized we need both later. I will updated it in newer commits

File: python/ray/train/batch_predictor.py
Comment (jiaodong) [reply python/ray/train/batch_predictor.py:182]: I think this is less desirable for sure, more like last step i put up to ensure it works before going to sleep lol.

From my observation disabling all casting leads to worse perf in this numpy first path too (more disk spilling at least), so we definitely want to ensure it's enabled throughout our execution. I think we just need to ensure the fall back is there on the final pandas casting, ideally without mutating anything global. I can give a try to the `cast_tensor_columns` flag.

But my question is i think `cast_tensor_columns` also has longer duration of the entire execution rather than last step, maybe the best we can do is to just leave it as is, only by adding a try-catch at the end ?

Comment (clarkzinzow) [reply python/ray/train/batch_predictor.py:182]: >I think we just need to ensure the fall back is there on the final pandas casting, ideally without mutating anything global. I can give a try to the cast_tensor_columns flag.

Agreed!

>But my question is i think cast_tensor_columns also has longer duration of the entire execution rather than last step, maybe the best we can do is to just leave it as is, only by adding a try-catch at the end ?

The local `cast_tensor_columns` variable (not the instance attribute) should only be applied to the model output cast below, AFAICT.

File: python/ray/train/batch_predictor.py
Comment (ericl) [inline python/ray/train/batch_predictor.py:178]: This is breaking abstraction boundaries. We should not check predictor. class identity within batch predictor.

File: python/ray/train/predictor.py
Comment (ericl) [inline python/ray/train/predictor.py:152]: We need to be checking if _predict_numpy/pandas/etc are defined for the given predictor (don't raise NotImplemented / are not equal to the base NotImplemented stub). Then, we should decide based on that and the data format the optimal one to call.

Btw, I noticed that we don't handle _predict_arrow here (the docs in this class seem to be wrong).

File: python/ray/train/batch_predictor.py
Comment (ericl) [reply python/ray/train/batch_predictor.py:182]: We shouldn't disable tensor casting automatically. That must be done by the user only.

Comment (jiaodong) [reply python/ray/train/batch_predictor.py:182]: Ah I see, we can leave the global and instance flag as is, only need to change `cast_tensor_columns` here for the numpy casting path

```
return convert_batch_type_to_pandas(
                    prediction_output, cast_tensor_columns
                )
```

File: python/ray/train/_internal/dl_predictor.py
Comment (ericl) [inline python/ray/train/_internal/dl_predictor.py:82]: This is pretty sketchy. Why not ask the user to batch in call_model instead? We shouldn't add "special-case magic conversions".

**Review by ericl — CHANGES_REQUESTED** [review]
Just did a quick review.  I have the following feedback:
1. It seems we are breaking abstraction boundaries in a few places.
2. The complexity of this change seems fairly large. We need to think about the design a little more to make sure the _predict_pandas/_predict_numpy/_predict_arrow fit together properly.
3. This change is much too large to merge for 2.0 at this point.

File: python/ray/train/_internal/dl_predictor.py
Comment (jiaodong) [reply python/ray/train/_internal/dl_predictor.py:82]: that could work too, user could do some rebatching with their `call_model()` implementation. I think we essentially just need to ensure we surface actionable errors for user to fix this, as our `_predict_pandas` also only covers a few cases only that will lead to errors for all model output formats 

```python
        # Handle model multi-output. For example if model outputs 2 images.
        if isinstance(output, dict):
            return pd.DataFrame(
                {k: TensorArray(self._tensor_to_array(v)) for k, v in output.items()}
            )
        else:
            return pd.DataFrame(
                {"predictions": TensorArray(self._tensor_to_array(output))},
                columns=["predictions"],
            )
```

**Comment by jiaodong** [comment]
> * The complexity of this change seems fairly large. We need to think about the design a little more to make sure the _predict_pandas/_predict_numpy/_predict_arrow fit together properly.

Agreed I think our current predictor code only works for a hardcoded pandas path with ~2 particular output shapes (`Tensor`, `Dict[str, Tensor]`) and everything else either breaks or expect user somehow to figure out missing pieces. I will put up a short design doc later for proper discussion around this that covers:

1) Support predict (pands / numpy / arrow) where we're able to automatically decide:
   - batch_format, data_format and prediction path used, ideally automatically 
2) How to achieve this with right level of abstractions and ordering so user knows what works out of the box
  - For example, if user brings TorchPredictor ideally we should be able to tell and adjust to numpy e2e automatically    

> * It seems we are breaking abstraction boundaries in a few places.

We need more design on this. It didn't looks like our current setup works beyond pandas and particularly not optimal for DL frameworks.

> * This change is much too large to merge for 2.0 at this point.

Then to unblock demo so we don't install from custom wheel, we need to revisit https://github.com/ray-project/ray/pull/27359/files to add a pandas fall through as smaller emergency exit. Let me iterate a bit on this today.

**Comment by jiaodong** [comment]
keep ailve

**Pull request closed without merging**
